### PR TITLE
Split communication logic from computation logic into orchestrator

### PR DESCRIPTION
**Pull request by fzyzcjy**
## Motivation

Currently, the managers do both communication (e.g. "send to another process using mq") and computation (e.g. "do the detokenization"). This PR separates them.

Moreover, by doing this refactor, adding SPMD logic would be a bit easier and cleaner.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html).

### CODE REVIEW

File: python/sglang/srt/orchestration/std/detokenizer.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import logging
```
Comment (ch-wan) [inline python/sglang/srt/orchestration/std/detokenizer.py:14]: copyright

File: python/sglang/srt/orchestration/std/launcher.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
+import asyncio
```
Comment (ch-wan) [inline python/sglang/srt/orchestration/std/launcher.py:14]: copyright

File: python/sglang/srt/orchestration/std/scheduler.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
+import faulthandler
```
Comment (ch-wan) [inline python/sglang/srt/orchestration/std/scheduler.py:14]: copyright

File: python/sglang/srt/managers/scheduler.py [hunk 4]
```diff
@@ -464,122 +396,59 @@ def watchdog_thread(self):
...
+            self.process_batch_result(batch, result)
+        else:
+            # When the server is idle, so self-check and re-init some states
+            self.check_memory()
+            self.new_token_ratio = self.init_new_token_ratio
 
-            self.last_batch = batch
+        self.last_batch = batch
 
     @torch.no_grad()
```
Comment (yangw1234) [inline python/sglang/srt/managers/scheduler.py]: A little concern on the usage of `@torch.no_grad()`.

In the original implementation, the `@torch.no_grad()` context manager is applied to the entire loop and is only entered and exited exactly once during the entire lifecycle; while in the new implementation the `@torch.no_grad()` context manager is entered and exited every step. 

So my question is that is there any additional overhead of repeatedly calling `@torch.no_grad()`?

Would it be better if we apply `with torch.no_grad()` in `run_scheduler_process`, e.g.

```python
with torch.no_grad():
    while True:
        communicator.recv_and_process_input_requests()
        scheduler.process_batch()
```

**Review by yangw1234 — APPROVED** [review]

**Comment by yangw1234** [comment]
LGTM except on minor comment.

File: python/sglang/srt/managers/scheduler.py (hunk 4 above)
Comment (fzyzcjy) [reply python/sglang/srt/managers/scheduler.py]: I also thought of this question when modifying it. The CI benchmark does seem to pass, but yes, I can move it to a `with` block outside the loop.

File: python/sglang/srt/orchestration/std/launcher.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        target=run_detokenizer_process,
+        args=(
+            server_args,
+            port_args,
+        ),
+    )
+    detoken_proc.start()
+
+    # Launch tokenizer process
+    orchestrator = StdOrchestrator(server_args, port_args)
```
Comment (shanyu-sys) [inline python/sglang/srt/orchestration/std/launcher.py]:
```suggestion
    # Launch orchestrator process
    orchestrator = StdOrchestrator(server_args, port_args)
```

**Review by shanyu-sys — APPROVED** [review]
Others LGTM

**Review by ch-wan — APPROVED** [review]

File: python/sglang/srt/managers/scheduler.py [hunk 6]
```diff
@@ -464,122 +396,57 @@ def watchdog_thread(self):
...
-                    self.tp_worker.cur_sampling_info if batch else None
-                )
-                self.process_batch_result(tmp_batch, tmp_result)
-            elif batch is None:
-                # When the server is idle, so self-check and re-init some states
-                self.check_memory()
-                self.new_token_ratio = self.init_new_token_ratio
-
-            self.last_batch = batch
+    def process_batch(self):
```
Comment (ch-wan) [inline python/sglang/srt/managers/scheduler.py]: What's the return value of process_batch? We need a docstring to describe it.

File: python/sglang/srt/managers/scheduler.py [hunk 7]
```diff
@@ -1493,7 +1363,7 @@ def stream_output(
 
             # Send to detokenizer
             if rids:
-                self.send_to_detokenizer.send_pyobj(
+                self.on_generation_output(
```
Comment (shanyu-sys) [inline python/sglang/srt/managers/scheduler.py:1399]: Shall we check `self.on_generation_output` before using it? e.g. Add an error message if `self.on_generation_output` is `None`.

Comment (fzyzcjy) [reply python/sglang/srt/managers/scheduler.py:1399]: It seems Scheduler class is an internal implementation detail and not user-facing API, so I personally feel both adding and not adding is OK.

**Comment by fzyzcjy** [comment]
After discussing with Lianmin, now I know there is no need to follow implementation requirements in e.g. 2736. Thus I spent several hours writing a new PR #3852, and this one is deprecated.
